POEM_056.md

```diff
@@ -0,0 +1,349 @@
+    d = a * b
+    e = c * 1.5
+    return d, e
+
+f = omf.wrap(func).defaults(shape=4, units='m')
+```
+
+### Getting the function default metadata
+
+Currently there doesn't seem to be a need to add this, as the defaults are automatically applied
+to the variable metadata, but a function to do this could easily be added later if needed.
+
+
+### Assumed default values
+
+In order to stay consistent with OpenMDAO's default value policy, we'll assume the same default
+behavior for functions, so if no shape or default value is supplied for a function variable, we'll
+assume that is has the value 1.0.  If the `shape` is provided and either the default value is
+not provided or is provided as a scalar value, then the assumed default value will be
+`np.ones(shape) * scalar_default_value`, where `scalar_default_value` is 1.0 if not specified.
+If `shape` is provided along with a non-scalar default value that has a different shape, then
+an exception will be raised.
+
+
+## Setting non-default function metadata
+
+The `metadata` decorator can be used to specify metadata that is intended to apply to the function 
+as a whole.  It's similar to `defaults` in that it describes metadata for the whole function, but unlike
+`defaults`, a conflicting metadata value will raise an exception.  One possible source of conflict 
+is if a function default argument differs in shape from a shape specified in `metadata`. For example:
+
+
+```python
+def func(a, b, c=np.ones(3)):  # shape of c is 3 so raise an exception because `metadata` shape is 4
+    d = a * b
+    e = c * 1.5
+    return d, e
+
+f = omf.wrap(func).metadata(shape=4, units='m')
+```
+
+
+## Variable names
+
+### Setting variable names
+
+We don't need to set input names because the function can always be inspected for those, but
+we also need to associate output names with function return values. Those return values, if they are 
+simple variables, for example, `return x, y`, will give us the output variable names we need.  
+But in those cases where the function returns expressions rather than simple variables, we need 
+another way to specify what the names of those output variables should be.  The `output_names` 
+method provides a concise way to do this, for example:
+
+
+```python
+def func(a, b, c):
+    return a * b * c, a * b -c
+
+f = omf.wrap(func).output_names('d', 'e')  # name of return values are 'd' and 'e'
+```
+
+If we don't want to bother with a separate decorator for output names, we could instead use the
+`add_outputs` decorator mentioned earlier, for example:
+
+```python
+def func(a, b, c):
+    return a * b * c, a * b -c
+
+# names of return values are 'd' and 'e' and they have no other metadata
+f = omf.wrap(func).add_outputs(d={}, e={})  
+```
+
+As mentioned above, if the function's return values are simple variable names, we don't need to
+call `output_names` because we can determine the names from inspecting the function, e.g., 
+
+
+```python
```

Comment:
It's a little unclear here if you're implying that you could skip the `add_input` and `add_output` calls all together here and just wrap the function. I think thats what you mean, but you should expand the example to clarify. 

can you call `add_input` without giving a name? will it follow the argument order in that case?

Reply:
We could definitely allow `add_input` to be called without providing a variable name since we can get the name by inspecting the function, but I'm not sure we want to.  It just seems like we'd be making it easier for a user to make a mistake.  We could potentially allow the same thing in the case of outputs if their name can be inferred from the source code, but I hesitate to allow that for the same reason.